src/api/procedures/setCustodian.ts

```diff
@@ -0,0 +1,108 @@
+import BigNumber from 'bignumber.js';
+
+import { Account, DefaultPortfolio, NumberedPortfolio } from '~/api/entities';
+import { PolymeshError, Procedure } from '~/base';
+import { AuthorizationType, ErrorCode } from '~/types';
+import { SignerType } from '~/types/internal';
+import {
+  authorizationToAuthorizationData,
+  dateToMoment,
+  signerToString,
+  signerValueToSignatory,
+} from '~/utils';
+
+export interface SetCustodianParams {
+  targetAccount: string | Account;
+  expiry?: Date;
+}
+
+/**
+ * @hidden
+ */
+export type Params = { did: string; id?: BigNumber } & SetCustodianParams;
+
+/**
+ * @hidden
+ */
+export async function prepareSetCustodian(
+  this: Procedure<Params, void>,
+  args: Params
+): Promise<void> {
+  const {
+    context: {
+      polymeshApi: {
+        tx: { identity },
+      },
+    },
+    context,
+  } = this;
+
+  const { targetAccount, expiry, did, id } = args;
+
+  const portfolio = id
+    ? new NumberedPortfolio({ did, id }, context)
+    : new DefaultPortfolio({ did }, context);
+
+  const [currentIdentity, custodian] = await Promise.all([
+    context.getCurrentIdentity(),
+    portfolio.getCustodian(),
+  ]);
+
+  if (custodian.did !== currentIdentity.did) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'You are not the custodian of this portfolio',
+    });
+  }
+
+  const address = signerToString(targetAccount);
+  const authorizationRequests = await currentIdentity.authorizations.getSent();
+
+  const hasPendingAuth = !!authorizationRequests.data.find(authorizationRequest => {
+    const {
+      target,
+      data,
+      data: { type },
+    } = authorizationRequest;
+
+    if (type === AuthorizationType.PortfolioCustody) {
+      const authorizationData = data as { value: NumberedPortfolio | DefaultPortfolio };
+
+      return (
+        signerToString(target) === address &&
```

Comment:
You won't need to check the target once you use `getReceived`, but you should check that the issuers are the same